### PR TITLE
Update k8s dependencies

### PR DESCRIPTION
**Pull request by ashwin-venkatesh**
Changes proposed in this PR:
Update client-go, apimachinery k8s-api and introduce controller runtime. This is the initial step required to introduce a controller.
Update the tests/code the had a change in the API signature.
Update CI golang version required by consul sdk 0.6.0 is 1.14 

How I've tested this PR:
Ran all the tests locally

How I expect reviewers to test this PR:
Things that used to work should continue to work

Checklist:
- [ ] Tests added
- [ ] CHANGELOG entry added (*HashiCorp engineers only, community PRs should not add a changelog entry*)

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,52 +1,41 @@
...
-	gopkg.in/alecthomas/kingpin.v2 v2.2.6 // indirect
-	k8s.io/api v0.0.0-20190325185214-7544f9db76f6
-	k8s.io/apimachinery v0.0.0-20190223001710-c182ff3b9841
-	k8s.io/client-go v8.0.0+incompatible
-	k8s.io/kube-openapi v0.0.0-20180731170545-e3762e86a74c // indirect
+	golang.org/x/xerrors v0.0.0-20200804184101-5ec99f83aff1 // indirect
+	k8s.io/api v0.18.2
+	k8s.io/apimachinery v0.18.2
+	k8s.io/client-go v0.18.2
+	sigs.k8s.io/controller-runtime v0.6.0
```
Comment (lkysow) [inline go.mod:38]: let's remove this

File: go.mod [hunk 2]
```diff
@@ -1,52 +1,41 @@
...
 )
 
 go 1.13
```
Comment (lkysow) [inline go.mod:41]:
```suggestion
go 1.14
```

File: catalog/to-consul/resource.go [hunk 3]
```diff
@@ -138,11 +139,11 @@ func (t *ServiceResource) Informer() cache.SharedIndexInformer {
 	return cache.NewSharedIndexInformer(
 		&cache.ListWatch{
 			ListFunc: func(options metav1.ListOptions) (runtime.Object, error) {
-				return t.Client.CoreV1().Services(metav1.NamespaceAll).List(options)
+				return t.Client.CoreV1().Services(metav1.NamespaceAll).List(context.Background(), options)
```
Comment (lkysow) [inline catalog/to-consul/resource.go:142]: I think we need to make these context.TODO() for non-tests

File: subcommand/acl-init/command.go [hunk 4]
```diff
@@ -127,7 +128,7 @@ func (c *Command) Run(args []string) int {
 }
 
 func (c *Command) getSecret(secretName string) (string, error) {
-	secret, err := c.k8sClient.CoreV1().Secrets(c.flagNamespace).Get(secretName, metav1.GetOptions{})
+	secret, err := c.k8sClient.CoreV1().Secrets(c.flagNamespace).Get(context.Background(), secretName, metav1.GetOptions{})
```
Comment (lkysow) [inline subcommand/acl-init/command.go:131]:
```suggestion
	secret, err := c.k8sClient.CoreV1().Secrets(c.flagNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
```

File: subcommand/create-federation-secret/command.go [hunk 5]
```diff
@@ -240,10 +241,10 @@ func (c *Command) Run(args []string) int {
 
 	// Now create the Kubernetes secret.
 	logger.Info("Creating/updating Kubernetes secret", "name", federationSecret.ObjectMeta.Name, "ns", c.flagK8sNamespace)
-	_, err = c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Create(federationSecret)
+	_, err = c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Create(context.Background(), federationSecret, metav1.CreateOptions{})
```
Comment (lkysow) [inline subcommand/create-federation-secret/command.go:244]:
```suggestion
	_, err = c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Create(context.TODO(), federationSecret, metav1.CreateOptions{})
```

File: subcommand/create-federation-secret/command.go [hunk 6]
```diff
@@ -240,10 +241,10 @@ func (c *Command) Run(args []string) int {
...
 	if k8serrors.IsAlreadyExists(err) {
 		logger.Info("Secret already exists, updating instead")
-		_, err = c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Update(federationSecret)
+		_, err = c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Update(context.Background(), federationSecret, metav1.UpdateOptions{})
```
Comment (lkysow) [inline subcommand/create-federation-secret/command.go:247]:
```suggestion
		_, err = c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Update(context.TODO(), federationSecret, metav1.UpdateOptions{})
```

File: subcommand/create-federation-secret/command.go [hunk 7]
```diff
@@ -297,7 +298,7 @@ func (c *Command) replicationToken(logger hclog.Logger) ([]byte, error) {
 	// This will run forever but it's running as a Helm hook so Helm will timeout
 	// after a configurable time period.
 	backoff.Retry(func() error {
-		secret, err := c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Get(secretName, metav1.GetOptions{})
+		secret, err := c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Get(context.Background(), secretName, metav1.GetOptions{})
```
Comment (lkysow) [inline subcommand/create-federation-secret/command.go:301]:
```suggestion
		secret, err := c.k8sClient.CoreV1().Secrets(c.flagK8sNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
```

File: subcommand/delete-completed-job/command.go [hunk 8]
```diff
@@ -102,7 +103,7 @@ func (c *Command) Run(args []string) int {
 	// Wait for job to complete.
 	logger.Info(fmt.Sprintf("waiting for job %q to complete successfully", jobName))
 	for {
-		job, err := c.k8sClient.BatchV1().Jobs(c.flagNamespace).Get(jobName, metav1.GetOptions{})
+		job, err := c.k8sClient.BatchV1().Jobs(c.flagNamespace).Get(context.Background(), jobName, metav1.GetOptions{})
```
Comment (lkysow) [inline subcommand/delete-completed-job/command.go:106]:
```suggestion
		job, err := c.k8sClient.BatchV1().Jobs(c.flagNamespace).Get(context.TODO(), jobName, metav1.GetOptions{})
```

File: subcommand/delete-completed-job/command.go [hunk 9]
```diff
@@ -141,7 +142,7 @@ func (c *Command) Run(args []string) int {
 	// ourselves.
 	logger.Info(fmt.Sprintf("job %q has succeeded, deleting", jobName))
 	propagationPolicy := metav1.DeletePropagationForeground
-	err = c.k8sClient.BatchV1().Jobs(c.flagNamespace).Delete(jobName, &metav1.DeleteOptions{
+	err = c.k8sClient.BatchV1().Jobs(c.flagNamespace).Delete(context.Background(), jobName, metav1.DeleteOptions{
```
Comment (lkysow) [inline subcommand/delete-completed-job/command.go:145]:
```suggestion
	err = c.k8sClient.BatchV1().Jobs(c.flagNamespace).Delete(context.TODO(), jobName, metav1.DeleteOptions{
```

File: subcommand/inject-connect/command.go [hunk 10]
```diff
@@ -379,12 +380,12 @@ func (c *Command) certWatcher(ctx context.Context, ch <-chan cert.Bundle, client
 
 			_, err := clientset.AdmissionregistrationV1beta1().
 				MutatingWebhookConfigurations().
-				Patch(c.flagAutoName, types.JSONPatchType, []byte(fmt.Sprintf(
+				Patch(context.Background(), c.flagAutoName, types.JSONPatchType, []byte(fmt.Sprintf(
```
Comment (lkysow) [inline subcommand/inject-connect/command.go:383]:
```suggestion
				Patch(context.TODO(), c.flagAutoName, types.JSONPatchType, []byte(fmt.Sprintf(
```

File: subcommand/server-acl-init/command.go [hunk 11]
```diff
@@ -644,7 +644,7 @@ func (c *Command) Run(args []string) int {
 // reading the Kubernetes Secret with name secretName.
 // If there is no bootstrap token yet, then it returns an empty string (not an error).
 func (c *Command) getBootstrapToken(secretName string) (string, error) {
-	secret, err := c.clientset.CoreV1().Secrets(c.flagK8sNamespace).Get(secretName, metav1.GetOptions{})
+	secret, err := c.clientset.CoreV1().Secrets(c.flagK8sNamespace).Get(context.Background(), secretName, metav1.GetOptions{})
```
Comment (lkysow) [inline subcommand/server-acl-init/command.go:647]:
```suggestion
	secret, err := c.clientset.CoreV1().Secrets(c.flagK8sNamespace).Get(context.TODO(), secretName, metav1.GetOptions{})
```

File: subcommand/service-address/command.go [hunk 12]
```diff
@@ -87,7 +88,7 @@ func (c *Command) Run(args []string) int {
 	var address string
 	var unretryableErr error
 	backoff.Retry(withErrLogger(log, func() error {
-		svc, err := c.k8sClient.CoreV1().Services(c.flagNamespace).Get(c.flagServiceName, metav1.GetOptions{})
+		svc, err := c.k8sClient.CoreV1().Services(c.flagNamespace).Get(context.Background(), c.flagServiceName, metav1.GetOptions{})
```
Comment (lkysow) [inline subcommand/service-address/command.go:91]:
```suggestion
		svc, err := c.k8sClient.CoreV1().Services(c.flagNamespace).Get(context.TODO(), c.flagServiceName, metav1.GetOptions{})
```

File: subcommand/acl-init/command.go (hunk 4 above)
Comment (ashwin-venkatesh) [reply subcommand/acl-init/command.go:131]: I did see those and skipped them intentionally. Given these subcommands are `main` files that we invoke Run(args) on, would we want to pass context into them in the future? I assumed they would behave provide the context themselves

Comment (lkysow) [reply subcommand/acl-init/command.go]: I figured we'd start a context.Background() at the top and then feed it through to these helpers.

**Pull request merged**